### ISSUE TITLE
What is the proper way to upload 12 files to S3?

### ISSUE THREAD

**Issue by Ricardo1980** [aws-amplify/aws-sdk-ios#2827]
Hello,

When using AWSS3TransferUtility.default().uploadData, what is the proper way to upload 12 files to S3?

Should I create my  own mechanism using GCD or NSOperationQueue? or should I AWSTask.init(forCompletionOfAllTasksWithResults:)?

What I want is to receive a callback once all files are uploaded or any of them is failing.

I don't mind if it is one after another or perhaps 2 or 3 at once. Maybe 12 at the same time is too much.

BTW, once I create a task of tasks with AWSTask, how do I run it?

Thanks a lot.

EDIT:
Reading this very old document, I wrote this code and it seems it is working:
https://aws.amazon.com/blogs/mobile/the-aws-mobile-sdk-for-ios-how-to-use-bftask/

```
func uploaImages(images: [Image], completion: @escaping (UploadImagesError?) -> Void) {

        guard let userIdentityId = AWSMobileClient.default().identityId else {
            completion(.noIdentityError)
            return
        }

        // upload all the pictures in serial
        var task = AWSTask<AnyObject>(result: nil)
        for image in images {
            task = task.continueWith { uploadTask -> AnyObject? in

                // if there is an error, stop the chain
                if let error = uploadTask.error {
                    print("--error uploading one picture: \(error)")
                    DispatchQueue.main.async {
                        completion(.serverError)
                    }
                    return nil
                }

                // next task
                print("--uploading a picture")
                let key = "protected/\(userIdentityId)/\(image.uuid)-large.jpg"
                return AWSS3TransferUtility.default().uploadData(image.data, key: key, contentType: "image/jpeg", expression: nil)
            }
        }

        // check the last task
        task.continueWith { uploadTask -> AnyObject? in
            DispatchQueue.main.async {
                if let error = uploadTask.error {
                    print("----there was an error uploading all pictures: \(error.localizedDescription)")
                    completion(.serverError)
                } else {
                    print("--all success")
                    completion(nil)
                }
            }
            return nil
        }
    }
```

What do you think about this approach?

Should I use
var task = AWSTask<AWSS3TransferUtilityUploadTask>(result: nil)
instead of 
var task = AWSTask<AnyObject>(result: nil)
?
If I do that, the generics in the continue closure is not working and I don't know how to fix it.

When there is an error, how can I extract the file that is failing? 

Thanks.

**Comment by lawmicha** [comment]
Hi @Ricardo1980, thanks for sharing the link and code snippet. I noticed that the `continueWith` block represents the upload task being created successfully or with an error. To handle the completion of the upload, you'll need to pass a completionHandler to `AWSS3TransferUtility.default().uploadData(completionHandler:)`. I imagine the tasks are creating successfully sequentially and the uploads are happening in parallel. 

If you define a completion handler closure like https://github.com/awslabs/aws-sdk-ios-samples/blob/main/S3TransferUtility-Sample/Swift/S3BackgroundTransferSampleSwift/UploadViewController.swift#L49

you should be able to access the `task` of type [AWSS3TransferUtilityTask](https://github.com/aws-amplify/aws-sdk-ios/blob/0f985322625566a566726e8ca2c24d5a5bc678b8/AWSS3/AWSS3TransferUtilityTasks.h#L96) and `error` and get the corresponding key

**Comment by Ricardo1980** [comment]
Hello @lawmicha 

There is something I don't understand.

What's the difference between this error
https://github.com/awslabs/aws-sdk-ios-samples/blob/523617320a5a7ba351549a7017a372f6ee2b3db3/S3TransferUtility-Sample/Swift/S3BackgroundTransferSampleSwift/DownloadViewController.swift#L82

and this error:
https://github.com/awslabs/aws-sdk-ios-samples/blob/523617320a5a7ba351549a7017a372f6ee2b3db3/S3TransferUtility-Sample/Swift/S3BackgroundTransferSampleSwift/DownloadViewController.swift#L64

I mean, is there something like an error related to the upload and another error related to the task? It is not clear. Can you help me with this? Thanks!

Right now my code is not working properly, because when the upload fails (for example using a bucket that does not exist or without permissions), I don't see that error in my code. I guess I have to fill the completion, but if I do that, howdo I stop the chain if there is an error?

What I want to do is upload several files in serial, if one fails, stop the chain and return error. If the last one success, the return success.

Thanks a lot for your time :)

**Comment by Ricardo1980** [comment]
Thanks @lawmicha 


But this is not clear.
It seems that to detect a correct upload I have to use the completion handler, with the point if using AWSTasks is to use continueWith and continueOnSuccessWith, so we can create rich logic like: try to upload these files in serial, if any of them fails, execute this.

But it appears that continueWith annd continueOnSuccessWith are executed all the time, when S3 is reached, even if the request fails.

Because of that, I don't really understand how can I send several tasks for serial execution. I mean, stop the execution if any of them fails and inform about that.

Can anyone explain to me how to do that? I don't find documentation about this.

Thanks a lot.

**Comment by lawmicha** [comment]
For TransferUtility, the continueWith is called immediately after the the AWSTask is created. The callback is successful if the storage operation was created successful and with an error if it failed and does not represent the actual upload over the network. If you start each AWSTask from a for loop, then all of the uploads will be occuring in parallel with each other since they are asynchronous. In the completion handler, if you want to stop execution due to an error, you will only be able to stop the ones that were not already completed. Attempting to stop one in progress may or may not be sucesssful but you can still try by retrieving your existing tasks and cancelling them. If you want a serial execution, you can start one upload and in the completion handler, start the second.

You can get existing upload tasks and cancel them like this
```
if let savedTasks = transferUtility!.getUploadTasks().result {
            for case let uploadtask as AWSS3TransferUtilityUploadTask in savedTasks {
                print("Cancelling upload task")
                uploadtask.cancel() 
            }
```

For the link, it looks like it must have been deleted from the [repo](https://github.com/aws-amplify/docs) and will need a bit more digging to pull it back up. I did find the content from this [wayback machine link](https://web.archive.org/web/20171210054441/http://docs.aws.amazon.com/mobile/sdkforios/developerguide/awstask.html)

**Comment by Ricardo1980** [comment]
@lawmicha 
From your link (and Bolts documentation too)
**"An AWSTask object represents the result of an asynchronous method. Using AWSTask, you can wait for an asynchronous method to return a value, and then do something with that value after it has returned. You can chain asynchronous requests instead of nesting them. This helps keep logic clean and code readable."**

From Bolts Documentation:
**"In the following example we want to fetch a user profile, then fetch a profile image, and if any of these operations failed - we still want to display an placeholder image:"**

```
fetchProfile(user).continueOnSuccessWithTask { task in
  return fetchProfileImage(task.result);
}.continueWith { task in
  if let image = task.result {
    return image
  }
  return ProfileImagePlaceholder()
}
```

Clearly, AWSS3TransferUtilityUploadTask is not following those ideas, and in fact not using Bolts/Promises power due to a wrong and poor implementation of AWSTasks. Whoever implemented AWSS3TransferUtility, didn't get the idea of Bolts (or Promises).

Do I open a bug?
Are you going to deprecate this and follow another approach? (I am asking this because maybe is not worth it to spend more time on this)

Thanks.

**Comment by lawmicha** [comment]
Hi @Ricardo1980, the latest guidance for new apps is to use [Amplify Storage](https://docs.amplify.aws/lib/storage/getting-started/q/platform/ios). This is the other approach that aims to replace TransferUtility. The programming model there is that the callback provides the completion, and return value is an operation that you can use to pause/resume/cancel the storage operation. Free free to open another issue over in the [Amplify iOS repo](https://github.com/aws-amplify/amplify-ios) if you have additional questions.

**Comment by Ricardo1980** [comment]
If someone is interested, this is what I am doing to upload several images in serial.

```
import Foundation
import AWSMobileClient
import AWSAppSync
import AWSS3
import CocoaLumberjack

struct Image {
    let uuid: String
    let data: Data
}

enum UploadImagesError: Error {
    case noImagesError
    case tooManyImagesError
    case repeatedNamesError
    case noIdentityError
    case serverError
}

final class Uploader {
    private var currentTask: AWSTask<AWSS3TransferUtilityUploadTask>?

    func uploadImages(images: [Image], totalUploadProgress: ((Double) -> Void)? = nil, completion: @escaping (UploadImagesError?) -> Void) {
        guard images.count != 0 else {
            completion(.noImagesError)
            return
        }

        guard images.count <= Constants.maxProfileImages else {
            completion(.noImagesError)
            return
        }

        guard images.count == Set(images.map { $0.uuid }).count else {
            completion(.repeatedNamesError)
            return
        }

        uploadNext(imageIndex: 0, images: images, totalUploadProgress: totalUploadProgress, completion: completion)
    }

    private func uploadNext(imageIndex: Int, images: [Image], totalUploadProgress: ((Double) -> Void)? = nil, completion: @escaping (UploadImagesError?) -> Void) {
        guard let userIdentityId = AWSMobileClient.default().identityId else {
            completion(.noIdentityError)
            clean()
            return
        }

        if imageIndex < images.count {
            let currentImage = images[imageIndex]
            let key = "protected/\(userIdentityId)/\(currentImage.uuid)-large.jpg"
            // let key = "pictures/\(userIdentityId)/\(currentImage.uuid)-large.jpg"

            let expression = AWSS3TransferUtilityUploadExpression()

            expression.progressBlock = { (task: AWSS3TransferUtilityTask, progress: Progress) in
                // DDLogVerbose("---progress: \(progress.fractionCompleted)")
                let progress = Double(imageIndex) * (1.0 / Double(images.count)) + (Double(progress.fractionCompleted) / Double(images.count))
                totalUploadProgress?(progress)
            }

            DDLogVerbose("Start upload")
            currentTask = AWSS3TransferUtility.default().uploadData(currentImage.data, key: key, contentType: "image/jpeg", expression: expression) { task, error in
                if let error = error {
                    DDLogVerbose("Error uploading picture: \(error.localizedDescription)")
                    completion(.serverError)
                    self.clean()
                } else {
                    DDLogVerbose("Picture uploaded.")
                    self.uploadNext(imageIndex: imageIndex + 1, images: images, totalUploadProgress: totalUploadProgress, completion: completion)
                }
            }

        } else {
            DDLogVerbose("No more images, all uploaded properly.")
            completion(nil)
            self.clean()
        }
    }

    private func clean() {
        currentTask?.result?.cancel()
        currentTask = nil
    }
}
```
